File: examples/python/phone_call.py
"""Single live phone call → DetectDeepfake.

A minimal worked example of the **integration pattern** for a real-time
audio source (FreeSWITCH `mod_audio_fork`, Twilio Media Streams, a SIPREC
tap, etc.) piped over a gRPC bidi stream. Reads a WAV file from disk and
streams its PCM in real time so the example is self-contained; in a
production integration you'd replace the WAV reader with your media-fork
socket reader — the rest of the file (bidi setup, paced send loop,
concurrent response reader) stays the same.

For load-testing N concurrent calls against a real backend (finding the
connection-count knee, comparing fp16/fp32 throughput, etc.), see the
sibling `phone_call_burst.py`.

CLI:
    python phone_call.py [--audio FILE] [--target localhost:50051]
                         [--chunk-ms 100] [--duration 30]
                         [--scenario-id ID]

Defaults:
  - if --audio is omitted, picks the first .wav in `examples/audio/`
    (gitignored — drop a fixture in).
"""

from __future__ import annotations

import argparse
import asyncio
from pathlib import Path

from aurigin.deepfake_detection.v1 import deepfake_detection_pb2 as pb
from aurigin.deepfake_detection.v1 import deepfake_detection_pb2_grpc as pb_grpc
from twilio.tme.extensions.common.v1 import audio_buffer_pb2 as ab_pb

from common import (
    ChunkRow, WavData, install_signal_shutdown, make_aio_channel, read_wav,
    transport_label,
)

DEFAULT_CHUNK_MS = 100
DEFAULT_DURATION_S = 30.0


def _resolve_audio(arg: Path | None) -> Path:
    if arg is not None:
        if not arg.is_file():
            raise FileNotFoundError(arg)
        return arg
    audio_dir = Path(__file__).resolve().parent.parent / "audio"
    wavs = sorted(audio_dir.glob("*.wav")) if audio_dir.is_dir() else []
    if not wavs:
        raise SystemExit(
            "No --audio supplied and no .wav files found in examples/audio/. "
            "Drop a 16-bit PCM or 32-bit IEEE-float WAV in examples/audio/ or pass one with --audio.",
        )
    return wavs[0]


async def send_call(call, wav: WavData, chunk_ms: int, duration_s: float) -> None:
    """Stream `wav` in real-time-paced AudioBuffer chunks until `duration_s` is up.

    Also imported by phone_call_burst.py — same loop, just instantiated N times.

    THIS LOOP IS THE INTEGRATION PATTERN. For a FreeSWITCH `mod_audio_fork`
    or Twilio Media Stream, replace the `pcm[cursor:…]` slicing with:

        async for frame in fork_socket:
            await call.write(pb.DetectDeepfakeRequest(
                audio=ab_pb.AudioBuffer(
                    type="audio/x-raw", format="S16LE",
                    channels=1, rate=8000,
                    duration_ns=int(len(frame) / 8000 * 1e9),
                    pts_ns=pts_ns,
                    size=len(frame), buffer=frame,
                ),
            ))
            pts_ns += duration_ns

    No manual `asyncio.sleep` pacing needed in that version — the socket IS
    the clock. We loop a finite WAV here just so the example self-contains.
    """
    bytes_per_chunk = max(1, int(wav.rate * chunk_ms / 1000) * wav.bytes_per_sample)
    chunk_seconds = chunk_ms / 1000

    await call.write(pb.DetectDeepfakeRequest(create_session_request=pb.CreateSessionRequest()))

    loop = asyncio.get_running_loop()
    next_send = loop.time()
    pts_ns = 0
    cursor = 0
    elapsed_s = 0.0

    while elapsed_s < duration_s:
        chunk = wav.samples[cursor : cursor + bytes_per_chunk]
        cursor += bytes_per_chunk
        if cursor >= len(wav.samples):
            cursor = 0  # loop the file so a short fixture can simulate a long call
        actual_frames = len(chunk) // wav.bytes_per_sample
        duration_ns = int(actual_frames / wav.rate * 1e9)

        await call.write(pb.DetectDeepfakeRequest(
            audio=ab_pb.AudioBuffer(
                type="audio/x-raw", format=wav.wire_format,
                channels=wav.channels, rate=wav.rate,
                duration_ns=duration_ns, pts_ns=pts_ns,
                size=len(chunk), buffer=chunk,
            ),
        ))
        pts_ns += duration_ns
        elapsed_s += chunk_seconds

        # Deadline-based wallclock pacing — ~1 s of audio per second of
        # wallclock. Drift-free: if a single send blocks longer than
        # chunk_seconds, the next sleep_for shrinks rather than compounding.
        next_send += chunk_seconds
        sleep_for = next_send - loop.time()
        if sleep_for > 0:
            await asyncio.sleep(sleep_for)

    await call.done_writing()


async def recv_call(
    call,
    *,
    label: str | None = None,
    sink: dict | None = None,
) -> None:
    """Print every server response as it arrives. Runs concurrently with the
    sender via asyncio.gather — that's the other half of the bidi pattern.

    Optional `label` / `sink` exist so phone_call_burst.py can reuse this
    same function: `label` prefixes every log line with the per-stream id,
    and `sink` (a dict) captures session_id + chunks + final aggregates so
    the caller can flush a CSV row block after the call ends.
    """
    def log(message: str) -> None:
        print(f"[{label}] {message}" if label else message)

    async for response in call:
        kind = response.WhichOneof("response")
        if kind == "create_session_response":
            session_id = response.create_session_response.session_id
            if sink is not None:
                sink["session_id"] = session_id
            log(f"📞 Session: {session_id}")
        elif kind == "analysis_result":
            r = response.analysis_result
            if sink is not None:
                sink["chunks"].append(ChunkRow(
                    offset_ms=r.audio_offset_ms, duration_ms=r.duration_ms,
                    confidence=r.confidence, label=r.label,
                ))
            log(
                f"   Analysis @ {r.audio_offset_ms / 1000:6.2f}s "
                f"| score={r.score:.3f} | label={r.label:18s} | confidence={r.confidence:.2f}",
            )
        elif kind == "final_result":
            f = response.final_result
            if sink is not None:
                sink["audio_duration_ms"] = f.total_audio_ms
                sink["global_result"] = f.overall_label
            log(
                f"☎️  Call ended | total={f.total_audio_ms / 1000:.2f}s "
                f"| score={f.overall_score:.3f} | label={f.overall_label} | analyses={f.analysis_count}",
            )


async def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__.split("\n", maxsplit=1)[0])
    parser.add_argument("--audio", type=Path, default=None, help="WAV file to stream (defaults to first in audio/)")
    parser.add_argument("--duration", type=float, default=DEFAULT_DURATION_S, help="Call length in seconds")
    parser.add_argument("--chunk-ms", type=int, default=DEFAULT_CHUNK_MS, help="Audio frame size in milliseconds")
    parser.add_argument("--target", default="localhost:50051", help="gRPC server host:port")
    parser.add_argument(
        "--scenario-id", default=None,
        help="Server-side simulator scenario to request (sent as x-scenario-id metadata).",
    )
    args = parser.parse_args()

    audio_path = _resolve_audio(args.audio)
    wav = read_wav(audio_path)
    scenario_suffix = f" | scenario={args.scenario_id}" if args.scenario_id else ""
    print(
        f"📞 Calling {args.target} | source={audio_path.name} "
        f"({wav.duration_s:.2f}s @ {wav.rate}Hz/{wav.channels}ch {wav.wire_format}) "
        f"| duration={args.duration:.1f}s | frame={args.chunk_ms}ms"
        f"{scenario_suffix} | transport={transport_label()}",
    )
    print("─" * 70)

    metadata: tuple[tuple[str, str], ...] = (
        (("x-scenario-id", args.scenario_id),) if args.scenario_id else ()
    )

    async with make_aio_channel(args.target) as channel:
        stub = pb_grpc.DeepfakeDetectionStub(channel)
        call = stub.DetectDeepfake(metadata=metadata) if metadata else stub.DetectDeepfake()

        # Send + receive concurrently: this is the bidi pattern. Sender writes
        # AudioBuffer messages at real-time pace; receiver reads
        # AnalysisResult / FinalResult messages as the server emits them.
        # Wrapped in an inner coroutine so we can asyncio.create_task() it —
        # gather() returns a _GatheringFuture (not a coroutine), which
        # create_task() rejects. The task wrapper is what
        # install_signal_shutdown cancels on Ctrl-C.
        async def _bidi() -> None:
            await asyncio.gather(send_call(call, wav, args.chunk_ms, args.duration), recv_call(call))

        task = asyncio.create_task(_bidi())
        install_signal_shutdown([task])
        try:
            await task
        except asyncio.CancelledError:
            pass  # graceful SIGINT/SIGTERM exit — banner already printed by the handler


def cli() -> None:
    """Sync entrypoint for `uv run phone-call`."""
    # See the long note in server.py: asyncio.run() installs a Python-level
    # SIGINT handler that conflicts with grpc.aio's own handlers, racing on
    # Ctrl-C and leaving a noisy traceback. Driving the loop manually keeps
    # the shutdown path clean: KeyboardInterrupt → finally → loop.close().
    loop = asyncio.new_event_loop()
    try:
        loop.run_until_complete(main())
    finally:
        loop.close()


if __name__ == "__main__":
    cli()

File: examples/typescript/phone_call.ts
// Single live phone call → DetectDeepfake.
//
// A minimal worked example of the **integration pattern** for a real-time
// audio source (FreeSWITCH `mod_audio_fork`, Twilio Media Streams, a SIPREC
// tap, etc.) piped over a gRPC bidi stream. Reads a WAV file from disk and
// streams its PCM in real time so the example is self-contained; in a
// production integration you'd replace the WAV reader with your media-fork
// socket reader — the rest of the file (bidi setup, paced send loop,
// concurrent response reader) stays the same.
//
// For load-testing N concurrent calls against a real backend (the
// recommended multi-call architecture — finding the connection-count
// knee, comparing fp16/fp32 throughput, etc.), see the sibling
// `phone_call_burst.ts`.
//
// CLI:
//   tsx phone_call.ts [--audio FILE] [--target localhost:50051]
//                     [--chunk-ms 100] [--duration 30]
//                     [--scenario-id ID]
//
// Defaults:
//   - if --audio is omitted, picks the first .wav in `examples/audio/`
//     (gitignored — drop a fixture in).

import * as fs from "node:fs";
import * as path from "node:path";
import { fileURLToPath } from "node:url";
import {
  Metadata,
  type ClientDuplexStream,
} from "@grpc/grpc-js";
import {
  DeepfakeDetectionClient,
  type DetectDeepfakeRequest,
  type DetectDeepfakeResponse,
} from "@aurigin/protos/aurigin/deepfake_detection/v1/deepfake_detection";

import {
  type ChunkRow,
  type WavData,
  channelCredentials,
  durationS as wavDurationS,
  installSignalShutdown,
  readWav,
  transportLabel,
} from "./common/index.js";

const DEFAULT_CHUNK_MS = 100;
const DEFAULT_DURATION_S = 30;

const __dirname = path.dirname(fileURLToPath(import.meta.url));

export type Call = ClientDuplexStream<DetectDeepfakeRequest, DetectDeepfakeResponse>;

export interface ResponseSink {
  sessionId: string;
  chunks: ChunkRow[];
  audioDurationMs: number;
  globalResult: string;
}

function resolveAudio(arg: string | null): string {
  if (arg) {
    if (!fs.existsSync(arg)) throw new Error(`No such file: ${arg}`);
    return arg;
  }
  const audioDir = path.join(__dirname, "..", "audio");
  if (!fs.existsSync(audioDir)) {
    throw new Error("No --audio supplied and examples/audio/ does not exist.");
  }
  const wavs = fs.readdirSync(audioDir).filter((f) => f.endsWith(".wav")).sort();
  if (wavs.length === 0) {
    throw new Error(
      "No --audio supplied and no .wav files found in examples/audio/. " +
        "Drop a 16-bit PCM or 32-bit IEEE-float WAV in examples/audio/ or pass one with --audio.",
    );
  }
  return path.join(audioDir, wavs[0]);
}

const sleep = (ms: number) => new Promise<void>((r) => setTimeout(r, ms));

// Stream `wav` in real-time-paced AudioBuffer chunks until `durationS` is up.
//
// Also imported by phone_call_burst.ts — same loop, just instantiated N times.
//
// THIS LOOP IS THE INTEGRATION PATTERN. For a FreeSWITCH `mod_audio_fork`
// or Twilio Media Stream, replace the `samples.subarray(cursor, …)` slicing
// with a `for await (const frame of forkSocket) { call.write(...) }` —
// no manual setTimeout pacing needed in that version, the socket IS the
// clock. We loop a finite WAV here just so the example self-contains.
export async function sendCall(
  call: Call, wav: WavData, chunkMs: number, durationS: number,
): Promise<void> {
  const bytesPerFrame = wav.bytesPerSample * wav.channels;
  const bytesPerChunk = Math.max(1, Math.floor((wav.rate * chunkMs) / 1000) * bytesPerFrame);
  const chunkS = chunkMs / 1000;

  call.write({ createSessionRequest: {} });

  let nextSend = Date.now();
  let ptsNs = 0n;
  let cursor = 0;
  let elapsedS = 0;

  while (elapsedS < durationS) {
    const end = Math.min(cursor + bytesPerChunk, wav.samples.length);
    const chunk = wav.samples.subarray(cursor, end);
    cursor = end;
    if (cursor >= wav.samples.length) cursor = 0;  // loop the file
    const actualFrames = chunk.length / bytesPerFrame;
    const durationNs = BigInt(Math.round((actualFrames / wav.rate) * 1e9));

    call.write({
      audio: {
        type: "audio/x-raw", format: wav.wireFormat,
        channels: wav.channels, rate: wav.rate,
        durationNs, ptsNs, size: BigInt(chunk.length), buffer: chunk,
      },
    });
    ptsNs += durationNs;
    elapsedS += chunkS;

    // Deadline-based wallclock pacing — drift-free: if a single write
    // blocks longer than chunkMs, the next sleepFor shrinks rather than
    // compounding.
    nextSend += chunkMs;
    const sleepFor = nextSend - Date.now();
    if (sleepFor > 0) await sleep(sleepFor);
  }

  call.end();
}

// Receive every server response as it arrives. Runs concurrently with the
// sender via Promise.all — that's the other half of the bidi pattern.
//
// Optional `label` / `sink` exist so phone_call_burst.ts can reuse this
// same function: `label` prefixes every log line with the per-stream id,
// and `sink` captures session_id + chunks + final aggregates so the
// caller can flush a CSV row block after the call ends.
export function recvCall(
  call: Call,
  opts: { label?: string; sink?: ResponseSink } = {},
): Promise<void> {
  const { label, sink } = opts;
  const log = (message: string): void => {
    console.log(label ? `[${label}] ${message}` : message);
  };
  return new Promise<void>((resolve, reject) => {
    call.on("data", (response: DetectDeepfakeResponse) => {
      if (response.createSessionResponse) {
        const sessionId = response.createSessionResponse.sessionId;
        if (sink) sink.sessionId = sessionId;
        log(`📞 Session: ${sessionId}`);
      } else if (response.analysisResult) {
        const r = response.analysisResult;
        if (sink) {
          sink.chunks.push({
            offsetMs: Number(r.audioOffsetMs),
            durationMs: Number(r.durationMs),
            confidence: r.confidence,
            label: r.label,
          });
        }
        const offsetS = (Number(r.audioOffsetMs) / 1000).toFixed(2).padStart(6);
        log(
          `   Analysis @ ${offsetS}s | score=${r.score.toFixed(3)} | ` +
            `label=${r.label.padEnd(18)} | confidence=${r.confidence.toFixed(2)}`,
        );
      } else if (response.finalResult) {
        const f = response.finalResult;
        if (sink) {
          sink.audioDurationMs = Number(f.totalAudioMs);
          sink.globalResult = f.overallLabel;
        }
        log(
          `☎️  Call ended | total=${(Number(f.totalAudioMs) / 1000).toFixed(2)}s | ` +
            `score=${f.overallScore.toFixed(3)} | label=${f.overallLabel} | ` +
            `analyses=${f.analysisCount}`,
        );
      }
    });
    call.on("end", resolve);
    call.on("error", reject);
  });
}

// ─── argv parser ────────────────────────────────────────────────────────────

interface Args {
  audio: string | null;
  duration: number;
  chunkMs: number;
  target: string;
  scenarioId: string | null;
}

function parseArgs(argv: string[]): Args {
  const out: Args = {
    audio: null,
    duration: DEFAULT_DURATION_S,
    chunkMs: DEFAULT_CHUNK_MS,
    target: "localhost:50051",
    scenarioId: null,
  };
  for (let i = 0; i < argv.length; i++) {
    const arg = argv[i];
    const next = () => argv[++i];
    switch (arg) {
      case "--audio": out.audio = next(); break;
      case "--duration": out.duration = Number(next()); break;
      case "--chunk-ms": out.chunkMs = Number(next()); break;
      case "--target": out.target = next(); break;
      case "--scenario-id": out.scenarioId = next(); break;
      case "-h": case "--help":
        console.log("Usage: tsx phone_call.ts [--audio FILE] [--duration SEC]");
        console.log("                         [--chunk-ms MS] [--target HOST:PORT]");
        console.log("                         [--scenario-id ID]");
        process.exit(0);
      default: throw new Error(`Unknown arg: ${arg}`);
    }
  }
  return out;
}

async function main() {
  const args = parseArgs(process.argv.slice(2));
  const audioPath = resolveAudio(args.audio);
  const wav = readWav(audioPath);

  const metadata = args.scenarioId ? new Metadata() : undefined;
  if (metadata && args.scenarioId) metadata.set("x-scenario-id", args.scenarioId);
  const scenarioSuffix = args.scenarioId ? ` | scenario=${args.scenarioId}` : "";

  console.log(
    `📞 Calling ${args.target} | source=${path.basename(audioPath)} ` +
      `(${wavDurationS(wav).toFixed(2)}s @ ${wav.rate}Hz/${wav.channels}ch ${wav.wireFormat}) ` +
      `| duration=${args.duration.toFixed(1)}s | frame=${args.chunkMs}ms` +
      `${scenarioSuffix} | transport=${transportLabel("client")}`,
  );
  console.log("─".repeat(70));

  const client = new DeepfakeDetectionClient(args.target, channelCredentials());
  try {
    const call: Call = metadata ? client.detectDeepfake(metadata) : client.detectDeepfake();

    // Graceful Ctrl-C: cancel the in-flight stream cleanly. installSignalShutdown
    // is the same helper phone_call_burst.ts uses for its N-call list.
    installSignalShutdown([call]);

    // Send + receive concurrently: this is the bidi pattern. sendCall
    // writes AudioBuffer messages at real-time pace; recvCall reads
    // AnalysisResult / FinalResult messages as the server emits them.
    await Promise.all([sendCall(call, wav, args.chunkMs, args.duration), recvCall(call)]);
  } finally {
    client.close();
  }
}

// Only run main() when this file is the process entry point. Without this
// guard, importing send_call / recv_call from phone_call_burst.ts would also
// trigger main() (the TS equivalent of Python's `if __name__ == "__main__":`).
if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch((err) => {
    console.error(err);
    process.exit(1);
  });
}
